crates/sqlbuiltins/src/functions/scalars/openai.rs

```diff
@@ -0,0 +1,245 @@
+            }
+            // openai_embed('api_key', '<model>', '<expr>')
+            3 => {
+                let creds = match args.get(0) {
+                    Some(Expr::Literal(ScalarValue::Utf8(v))) => {
+                        Some(v.clone().unwrap().to_string())
+                    }
+                    Some(Expr::ScalarVariable(_, values)) => {
+                        let scalar = creds_from_arg(values.clone());
+                        scalar
+                    }
+                    _ => return Err(DataFusionError::Plan("Invalid argument".to_string())),
+                };
+
+                let model = model_from_arg(&args[1])?;
+                (creds, model, 2)
+            }
+            _ => return Err(DataFusionError::Plan("Invalid argument count".to_string())),
+        };
+        let Some(creds) = creds else {
+            return Err(DataFusionError::Plan(
+                "No API key or credential provided".to_string(),
+            ));
+        };
+
+        let model_len = model.len();
+        let return_type_fn: ReturnTypeFunction = Arc::new(move |_| {
+            let f = Field::new("item", DataType::Float32, true);
+            let dtype = DataType::FixedSizeList(Arc::new(f), model_len);
+            Ok(Arc::new(dtype))
+        });
+
+        let scalar_fn_impl: ScalarFunctionImplementation = Arc::new(move |args| {
+            let input = match &args[0] {
+                ColumnarValue::Array(arr) => match arr.data_type() {
+                    DataType::Utf8 => EmbeddingInput::StringArray(
+                        arr.as_string::<i32>()
+                            .into_iter()
+                            .map(|s| s.unwrap_or_default().to_string())
+                            .collect(),
+                    ),
+                    DataType::LargeUtf8 => EmbeddingInput::StringArray(
+                        arr.as_string::<i64>()
+                            .into_iter()
+                            .map(|s| s.unwrap_or_default().to_string())
+                            .collect(),
+                    ),
+                    _ => return Err(DataFusionError::Plan("Invalid argument".to_string())),
+                },
+                ColumnarValue::Scalar(ScalarValue::Utf8(v) | ScalarValue::LargeUtf8(v)) => {
+                    EmbeddingInput::StringArray(vec![v.clone().unwrap_or_default()])
+                }
+                _ => return Err(DataFusionError::Plan("Invalid argument".to_string())),
+            };
+
+            let client = Client::with_config(OpenAIConfig::new().with_api_key(creds.clone()));
+            let embed = client.embeddings();
+            let res = embed.create(CreateEmbeddingRequest {
+                model: model.to_string(),
+                input: input,
+                encoding_format: Some(EncodingFormat::Float),
+                user: None,
+                dimensions: None,
+            });
+
+            // no way around blocking here. Expressions are not async
```

Comment:
seems like a user could easily use this function to take down a node... might want/need to feature flag it into only the bindings?

Is there *any* way to avoid this even if it's gross or a lot of work? not necessarily to do it, but just to understand the decision better.

Reply:
No, expressions are not async, so we pretty much have to block on this. AFAIK, there is no way around this without rewriting datafusion exprs. 

IMO, we're better off doing some rate limiting, throttling, and aggressive timeouts than to try and avoid blocking here. I agree that it's not ideal, but there isn't much we can do.

Reply:
I definitely think if we do this implementation that it needs to have _some_ kind limitation on it so that it won't bring down cloud or allow one user to DoS a cloud node, or we should feature flag it so it only exists in local mode. 

Can't execution plans and table/funcs be implemented as streams which are async? Is there a way to have a scalar function call a table function, or implement this as a table function? (mostly just spitballing).

Reply:
Datafusion async UDF issue: https://github.com/apache/arrow-datafusion/issues/6518

I think given the current limitation, a table func would be better here. It would also provide the option of returning a "column" instead of an array which might make it easier to run math functions on (I have no evidence on this last part).

Reply:
I'm still not sure we should flat out forbid using this technique in functions. After all, when we eventually support UDF's we'll have no control over what those exprs are doing. 

This is much more controlled. OpenAI already has pretty aggressive [rate limits](https://platform.openai.com/docs/guides/rate-limits?context=tier-free), and the need for a user provided credential would likely result in rate limits from OpenAI well before anything negative would happen to our cloud infra. 

While I do think this is a valid concern, I don't think the risk surface area is that large either. 

I'd much prefer keeping this as an expression instead of trying to force it into something it's not _(table function)_.

Reply:
I also think the idea of feature flagging is somewhat appealing here. We could feature flag this out of the shared cloud instances & reserve it for embedded, self-deployed, or a dedicated cloud instance.

Reply:
I'd prefer this not being feature flagged. Having feature flags combined with our current configuration will make it easier to miss things during testing.